auth_dynamic_groups_relation/models/res_groups.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2019 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+# pylint: disable=missing-docstring,invalid-name
+from odoo import _, api, fields, models
+
+COUNT_RELATION_STATEMENT = """\
+SELECT COUNT(*) AS kount
+ FROM dynamic_group_relation_type_rel rel
+ JOIN res_partner_relation_all AS rpra
+   ON rel.type_selection_id = rpra.type_selection_id
+ WHERE rel.gid = %(group_id)s
+   AND (rpra.this_partner_id = %(partner_id)s OR rpra.other_partner_id = %(partner_id)s)
+   AND (rpra.date_start IS NULL or rpra.date_start <= CURRENT_DATE)
+   AND (rpra.date_end IS NULL or rpra.date_end >= CURRENT_DATE)
+"""
+
+
+class ResGroups(models.Model):
+    _inherit = 'res.groups'
+
+    dynamic_method_relation = fields.Selection(
+        selection="_get_group_use_selection",
+        string="Use partner relations",
+        help="Include partners that have a specific type of relation")
+    relation_type_ids = fields.Many2many(
+        comodel_name='res.partner.relation.type.selection',
+        relation='dynamic_group_relation_type_rel',
+        column1='gid', column2='type_selection_id',
+        string="Relation types that grant group membership",
+        help="Specify the relations types that, if a user is connected"
+             " with another partner through such a relation, grants the"
+             "  user membership of this group.")
+    allow_companies = fields.Boolean(
+        help="Normally users should be persons, and therefore only the"
+             " side of relations that apply to persons should be selectable.\n"
+             "Allowing companies as users, will make all relations"
+             " available for selection.")
+
+    @api.onchange('allow_companies')
+    def _onchange_allow_companies(self):
+        self.ensure_one()
+        relation_type_domain = [(1, '=', 1)]
+        if not self.allow_companies:
+            relation_type_domain = [
+                '|',
+                ('contact_type_this', '=', 'p'),
+                ('contact_type_this', '=', False)]
+        return {'domain': {'relation_type_ids': relation_type_domain}}
+
+    @api.multi
+    def dynamic_method_relation_check(self, user):
+        """Check wether user has an active relation that gives group access."""
+        self.ensure_one()
+        self.env.cr.execute(
+            COUNT_RELATION_STATEMENT,
+            {
+                'group_id': self.id,
+                'partner_id': user.partner_id.id,
+            }
+        )
+        kount = self.env.cr.fetchone()[0]
+        message_parms = {
+            'group': '%d:%s' % (self.id, self.display_name),
+            'user': '%d:%s' % (user.id, user.display_name),
+            'kount': kount,
+        }
+        self._debug_message(
+            _("User %(user)s has %(kount)d relations qualifying"
+              " for dynamic group %(group)s"),
+            message_parms)
+        return kount > 0
```

Comment:
Since partner relations can also be queried using [regular Odoo domain notation](https://github.com/OCA/partner-contact/blob/11.0/partner_multi_relation/tests/test_partner_relation.py#L248), wouldn't it be easier just to allow to specify a "domain" that defines a dynamic group, in addition to a Python formula. It seems like a simpler and more flexible change.

Reply:
While that would be more flexible, it would also be more complicated for the end-user. This was for a project that has been stopped unfortunately, but still I think just being able to put board members (people that have the relation 'is a board member') into a group giving access to board member type functions - of course this is just an example-, is a rather straight forward functionality.

Reply:
Fair enough, i think it's an and/and thing tho - supporting domain notation would be useful as well for speed of evaluation.